packages/backend/src/modules/finality/FinalityIndexer.ts

```diff
@@ -0,0 +1,184 @@
+import { Logger } from '@l2beat/backend-tools'
+import { assert, UnixTime } from '@l2beat/shared-pure'
+import { ChildIndexer } from '@l2beat/uif'
+
+import { IndexerStateRepository } from '../../peripherals/database/repositories/IndexerStateRepository'
+import { LivenessIndexer } from '../liveness/LivenessIndexer'
+import {
+  FinalityRecord,
+  FinalityRepository,
+} from './repositories/FinalityRepository'
+import { FinalityConfig } from './types/FinalityConfig'
+import { findConfigurationsToSync } from './utils/findConfigurationsToSync'
+
+/*
+  Once per day we want to fetch finality data for each project for last 24h, with granularity of 10 minutes,
+  so in every hour there will be 6 calls
+*/
+const FINALITY_GRANULARITY = 24 * 6
+
+export class FinalityIndexer extends ChildIndexer {
+  readonly indexerId = 'finality_indexer'
+  readonly minTimestamp = UnixTime.fromDate(
+    new Date('2024-02-07T00:00:00.000Z'),
+  )
+
+  constructor(
+    logger: Logger,
+    parentIndexer: LivenessIndexer,
+    private readonly stateRepository: IndexerStateRepository,
+    private readonly finalityRepository: FinalityRepository,
+    private readonly runtimeConfigurations: FinalityConfig[],
+  ) {
+    super(logger, [parentIndexer])
+  }
+
+  override async start(): Promise<void> {
+    this.logger.info('Starting...')
+    await this.initialize()
+    await super.start()
+  }
+
+  override async update(from: number, to: number): Promise<number> {
+    const targetTo = new UnixTime(from)
+      .toStartOf('day')
+      .add(1, 'days')
+      .toNumber()
+
+    if (targetTo < this.minTimestamp.toNumber()) {
+      this.logger.debug(
+        'Update skipped: target earlier than minimumTimestamp',
+        { from, targetTo },
+      )
+      return targetTo
+    }
+
+    if (targetTo > UnixTime.now().toNumber()) {
+      return to
+    }
+
+    const targetTimestamp = new UnixTime(targetTo)
+
+    const configurationsToSync = await this.getSyncStatus(targetTimestamp)
+
+    if (configurationsToSync.length === 0) {
+      this.logger.debug('Update skipped: no configurations to sync', {
+        from,
+        to: targetTo,
+      })
+      return targetTimestamp.toNumber()
```

Comment:
Watch out here. This can result in `targetTimestamp > to` being returned. This will never happen in our current setup, but a change in a parentIndexer for this one can become a problem.